internal/controller/environment_controller.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	etosv1alpha1 "github.com/eiffel-community/etos/api/v1alpha1"
+)
+
+var (
+	typeReady     = "Ready"
+	typeAvailable = "Available"
+	checkInterval = 30 * time.Second
+)
+
+// EnvironmentReconciler reconciles a Environment object
+type EnvironmentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments/finalizers,verbs=update
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=providers,verbs=get
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=providers/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *EnvironmentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Get environment if exists.
+	environment := &etosv1alpha1.Environment{}
+	err := r.Get(ctx, req.NamespacedName, environment)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("environment not found. ignoring object")
```

Comment:
General question: is there enough info logging? I guess observability is important for controllers in general, we want to be able to see the state.

Reply:
There's never enough logging :) For now I think it's good enough. We need to identify which information is important to log and then make sure we log enough, but not too much.
For the line that you posted this comment, that is a message that gets logged if the `Environment` is deleted, I.e. using `kubectl delete environment/name`.